### PR TITLE
Set extension properties directly

### LINKED ISSUE #41101: Quarkus logs warning for properties set by Observability Dev Services with Grafana OTEL LGTM
### Describe the bug

I use Observability Dev Services with Grafana OTEL LGTM and see reports about unrecognized properties I didn't set.

### Expected behavior

No warning unless I can make action that prevents warning.

### Actual behavior

```
2024-06-10 20:02:37,481 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.otel-collector.url" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2024-06-10 20:02:37,481 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.grafana.url" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
is proxy options enabled? false
```

### How to Reproduce?

1. in Quarkus main run `mvn clean test -f integration-tests/observability-lgtm/ `

### Output of `uname -a` or `ver`

Fedora 38

### Output of `java -version`

21

### Quarkus version or git rev

999-SNAPSHOT

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.4

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by brunobat**
fixes #41101 

- Detect which extensions (Otel, Micrometer OTLP registry) are being used and set properties accordingly. 
- Remove Quarkus prefix to service properties because they will not be used by the users.
- Update documentation
- Boy Scout stuff like removing lambda expressions from extension and use inner classes for test profiles.

**There are breaking changes.**
The following properties are removed because they are not needed anymore:
- `quarkus.otel-collector.url`
- `quarkus.grafana.url`

The extension will now detect if Micrometer OTLP registry or the OpenTelemetry extension are present and set the relevant properties for the user.

### CODE REVIEW

File: extensions/observability-devservices/deployment/src/main/java/io/quarkus/observability/deployment/ObservabilityDevServiceProcessor.java [hunk 1]
```diff
@@ -152,19 +152,22 @@ public void startContainers(LaunchModeBuildItem launchMode,
...
 
             if (firstStart.computeIfAbsent(devId, x -> true)) {
-                Runnable closeTask = () -> {
-                    DevServicesResultBuildItem.RunningDevService current = devServices.get(devId);
-                    if (current != null) {
-                        try {
-                            current.close();
-                        } catch (Throwable t) {
-                            log.errorf("Failed to stop %s container", devId, t);
+                Runnable closeTask = new Runnable() {
```
Comment (alesj) [inline extensions/observability-devservices/deployment/src/main/java/io/quarkus/observability/deployment/ObservabilityDevServiceProcessor.java:155]: Just leave the lambda?

File: integration-tests/observability-lgtm/src/test/java/io/quarkus/observability/test/LgtmTestBase.java [hunk 2]
```diff
@@ -13,15 +13,15 @@
...
+    @ConfigProperty(name = "grafana.endpoint")
+    String endpoint;
 
     @Test
     public void testTracing() {
         log.info("Testing Grafana ...");
         String response = RestAssured.get("/api/poke?f=100").body().asString();
         log.info("Response: " + response);
-        GrafanaClient client = new GrafanaClient("http://" + url, "admin", "admin");
+        GrafanaClient client = new GrafanaClient(endpoint, "admin", "admin");
```
Comment (alesj) [inline integration-tests/observability-lgtm/src/test/java/io/quarkus/observability/test/LgtmTestBase.java:24]: Does this work w/o `http://` ?

Comment (brunobat) [reply integration-tests/observability-lgtm/src/test/java/io/quarkus/observability/test/LgtmTestBase.java:24]: Not sure, I'm just trying to make things compatible with the previous version.

File: extensions/observability-devservices/deployment/src/main/java/io/quarkus/observability/deployment/ObservabilityDevServiceProcessor.java (hunk 1 above)
Comment (brunobat) [reply extensions/observability-devservices/deployment/src/main/java/io/quarkus/observability/deployment/ObservabilityDevServiceProcessor.java:155]: Nope. This is a good practice in the extensions to reduce the number of classes at runtime. We don't care about it on user code but extensions should.

**Comment by brunobat** [comment]
@alesj @melloware The PR is ready for review/

**Review by melloware — APPROVED** [review]
LGTM

File: extensions/observability-devservices/testlibs/devresource-common/src/main/java/io/quarkus/observability/devresource/DevResourceLifecycleManager.java [hunk 3]
```diff
@@ -28,8 +28,18 @@ public interface DevResourceLifecycleManager<T extends ContainerConfig> extends
...
     T config(ModulesConfiguration configuration);
 
+    /**
+     * Get resource's config from main observability configuration and extension catalog
+     *
+     * @param configuration main observability configuration
+     * @param catalog observability catalog. If OpenTelemetry or Micrometer are enabled.
+     * @return module's config
+     */
+    T config(ModulesConfiguration configuration, ExtensionsCatalog catalog);
```
Comment (alesj) [inline extensions/observability-devservices/testlibs/devresource-common/src/main/java/io/quarkus/observability/devresource/DevResourceLifecycleManager.java:41]: Have this one call deprecated by default -- so we don't break existing impls (if there are any ...).

**Comment by alesj** [comment]
And perhaps add prop names changes to CLI `update command?

**Review by alesj — APPROVED** [review]

**Comment by alesj** [comment]
> And perhaps add prop names changes to CLI `update command?

... and we definitely need to mention the name changes in the release notes ...

**Comment by brunobat** [comment]
> e definitely need to mention the name changes in the release notes ...

I will. I'm not sure about the update command, can you please do it later?

**Review by melloware — APPROVED** [review]
this looks great

**Comment by gsmet** [comment]
Hmmm, why are we having properties outside of the `quarkus.` namespace? We worked hard to normalize it, I'm not exactly excited by the idea of not following this convention.

**Comment by melloware** [comment]
@gsmet i think the properties in question can't be changed by user they are dynamically allocated by the system right?

**Comment by brunobat** [comment]
correct @melloware.
The property values are generated by the extension and should be read only.
In normal cases the user don't even need them.

**Comment by geoand** [comment]
Even in light of that explanation, I still don't understand why the properties are not under the Quarkus namespace. 
Also, what does "read only" mean? IIUC from the docs, they are settable by the user aren't they?

**Comment by geoand** [comment]
> If we place them under the quarkus namespace I have to create configurations for them, to remove the warning, and to make sure the user doesn't configure them, there are ways but it's not strait-forward and all this sounds overkill for properties that shouldn't even be public.

Sure, but this is the same argument that people have always used and we always push back against 😉

**Comment by brunobat** [comment]
Ok, then I will just say the properties were removed and are not available anymore.

**Comment by geoand** [comment]
👍🏼

**Comment by brunobat** [comment]
@gsmet removed references to those properties.

**Review by melloware — APPROVED** [review]

**Comment by maxandersen** [comment]
> Remove Quarkus prefix to service properties because they will not be used by the users.

makes me squint seeing we remove quarkus prefix...aren't that prefix used for various config setup? ...is this because these properties are truly devmode internal only / used to decouple different services? should we maybe have another prefix for them or?

**Comment by brunobat** [comment]
> > Remove Quarkus prefix to service properties because they will not be used by the users.
> 
> makes me squint seeing we remove quarkus prefix...aren't that prefix used for various config setup? ...is this because these properties are truly devmode internal only / used to decouple different services? should we maybe have another prefix for them or?

Yes @maxandersen , they are internal properties that should not be changed by users and should not be documented. Therefore, the missing `quarkus` prefix because as soon as we add it, we need to create a property to not have a console warning and creating a property requires documentation. 
As far as I know, this is the first time we have "read-only" properties like these ones. It would be nice to have a different prefix and behaviour for them. What do you think, @radcortez ?

**Comment by radcortez** [comment]
> As far as I know, this is the first time we have "read-only" properties like these ones. It would be nice to have a different prefix and behaviour for them. What do you think, @radcortez ?

You can add such properties as `Map<String, String>` to a `BUILD_AND_RUN_TIME_FIXED`. Just pick a path for the `Map`. and then the validation will mark as valid any property name as part of the `Map`. It still requires documentation, but you can add something generic `Internal Quarkus container properties... etc.`. 

Yes, users could still set the configuration from build time, but since you can always override it, disregarding the one set by the user. When you get to dev mode, because the phase is `BUILD_AND_RUN_TIME_FIXED`, it cannot be changed.

**Comment by brunobat** [comment]
> > As far as I know, this is the first time we have "read-only" properties like these ones. It would be nice to have a different prefix and behaviour for them. What do you think, @radcortez ?
> 
> You can add such properties as `Map<String, String>` to a `BUILD_AND_RUN_TIME_FIXED`. Just pick a path for the `Map`. and then the validation will mark as valid any property name as part of the `Map`. It still requires documentation, but you can add something generic `Internal Quarkus container properties... etc.`.
> 
> Yes, users could still set the configuration from build time, but since you can always override it, disregarding the one set by the user. When you get to dev mode, because the phase is `BUILD_AND_RUN_TIME_FIXED`, it cannot be changed.

That's why I prefer not even document the properties... Too much work for nothing. :)

File: integration-tests/observability-lgtm/src/test/java/io/quarkus/observability/test/LgtmTestBase.java (hunk 2 above)
Comment (alesj) [reply integration-tests/observability-lgtm/src/test/java/io/quarkus/observability/test/LgtmTestBase.java:24]: Did you check / fix this?

File: extensions/observability-devservices/testlibs/devresource-common/src/main/java/io/quarkus/observability/devresource/DevResourceLifecycleManager.java (hunk 3 above)
Comment (alesj) [reply extensions/observability-devservices/testlibs/devresource-common/src/main/java/io/quarkus/observability/devresource/DevResourceLifecycleManager.java:41]: Not this one -- deprecate the old one.
And have the new one call the old one by default.

**Comment by alesj** [comment]
@brunobat apart from those comments I just made, lgtm.

**Review by alesj — APPROVED** [review]

File: integration-tests/observability-lgtm/src/test/java/io/quarkus/observability/test/LgtmTestBase.java (hunk 2 above)
Comment (brunobat) [reply integration-tests/observability-lgtm/src/test/java/io/quarkus/observability/test/LgtmTestBase.java:24]: the endpoint already includes the `http://` part

File: extensions/observability-devservices/testlibs/devresource-common/src/main/java/io/quarkus/observability/devresource/DevResourceLifecycleManager.java (hunk 3 above)
Comment (brunobat) [reply extensions/observability-devservices/testlibs/devresource-common/src/main/java/io/quarkus/observability/devresource/DevResourceLifecycleManager.java:41]: I thought about it, but they are needed for different things and I didn't want to mess things too much.

Comment (alesj) [reply extensions/observability-devservices/testlibs/devresource-common/src/main/java/io/quarkus/observability/devresource/DevResourceLifecycleManager.java:41]: True ... but this way we can don't break any existing one (if there are some :-) ...

**Review by gsmet — APPROVED** [review]
Ok, let's make progress on this, we can still tweak things later if needed.

**Pull request merged**